src/Mono.Android/Microsoft.Android.Runtime/ManagedValueManager.cs

```diff
@@ -247,21 +278,163 @@ Type GetDeclaringType (ConstructorInfo cinfo) =>
 
 	public override List<JniSurfacedPeerInfo> GetSurfacedPeers ()
 	{
-		if (RegisteredInstances == null)
-			throw new ObjectDisposedException (nameof (ManagedValueManager));
+		ThrowIfDisposed ();
+
+		// Remove any collected contexts before iterating over all the registered instances
+		CollectPeers ();
 
 		lock (RegisteredInstances) {
 			var peers = new List<JniSurfacedPeerInfo> (RegisteredInstances.Count);
 			foreach (var e in RegisteredInstances) {
 				foreach (var p in e.Value) {
-					peers.Add (new JniSurfacedPeerInfo (e.Key, new WeakReference<IJavaPeerable> (p)));
+					if (p.Target is IJavaPeerable peer) {
+						peers.Add (new JniSurfacedPeerInfo (e.Key, new WeakReference<IJavaPeerable> (peer)));
+					}
 				}
 			}
 			return peers;
 		}
 	}
 
-	const   BindingFlags    ActivationConstructorBindingFlags   = BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance;
+	unsafe struct ReferenceTrackingHandle : IDisposable
+	{
+		private WeakReference<IJavaPeerable> _weakReference;
+		private HandleContext* _context;
```

Comment:
Members are private by default

```suggestion
		WeakReference<IJavaPeerable> _weakReference;
		HandleContext* _context;
```